src/envoy/tcp/alpn_proxy/alpn_proxy.cc

```diff
@@ -0,0 +1,164 @@
+      reinterpret_cast<const char*>(&initial_header), sizeof(initial_header))};
+  auto proxy_header_buffer = std::make_unique<::Envoy::Buffer::OwnedImpl>(
+      proxy_header.SerializeAsString());
+  proxy_header_buffer->prepend(initial_header_buffer);
+  return proxy_header_buffer;
+}
+
+}  // namespace
+
+AlpnProxyConfig::AlpnProxyConfig(
+    const std::string& stat_prefix, const std::string& protocol,
+    const std::string& node_metadata_id,
+    const envoy::tcp::alpnproxy::config::AlpnProxy::FilterDirection&
+        filter_direction,
+    Stats::Scope& scope)
+    : scope_(scope),
+      stat_prefix_(stat_prefix),
+      protocol_(protocol),
+      node_metadata_id_(node_metadata_id),
+      filter_direction_(filter_direction),
+      stats_(generateStats(stat_prefix, scope)) {}
+
+Network::FilterStatus AlpnProxyFilter::onData(Buffer::Instance& data, bool) {
+  // No work needed if we have read connection protocol and its not the expected
+  // alpn protocol or if alpn protocol is found but we have yet not consumed
+  // metadata.
+  if ((connection_protocol_read_ && !alpn_protocol_found_) ||
+      (alpn_protocol_found_ && !done_on_data_)) {
+    return Network::FilterStatus::Continue;
+  }
+  if (read_callbacks_->connection().nextProtocol() != config_->protocol_) {
+    connection_protocol_read_ = true;
+    config_->stats().alpn_protocol_not_found_.inc();
+    return Network::FilterStatus::Continue;
+  }
+  // Write metadta first time protocol is found.
+  if (!alpn_protocol_found_) {
+    config_->stats().alpn_protocol_found_.inc();
+    alpn_protocol_found_ = true;
+    if (local_info_.node().has_metadata()) {
+      auto metadata_fields = local_info_.node().metadata().fields();
+      auto node_metadata = metadata_fields.find(config_->node_metadata_id_);
+      if (node_metadata != metadata_fields.end()) {
+        std::unique_ptr<::Envoy::Buffer::OwnedImpl> buf =
+            ConstructProxyHeaderData(node_metadata->second.struct_value());
+        write_callbacks_->injectWriteDataToFilterChain(*buf, false);
+
+        read_callbacks_->connection().streamInfo().setDynamicMetadata(
+            downstreamDynamicDataKey, node_metadata->second.struct_value());
+        config_->stats().metadata_added_.inc();
+      }
+    }
+  }
+
+  // Consume Metadata in Downstream Filter.
+  if (config_->filter_direction_ ==
+      envoy::tcp::alpnproxy::config::AlpnProxy_FilterDirection::
+          AlpnProxy_FilterDirection_DOWNSTREAM) {
+    config_->stats().downstream_filter_found_.inc();
+    const uint32_t initial_header_length = sizeof(AlpnProxyInitialHeader);
+    if (data.length() < initial_header_length) {
+      config_->stats().initial_header_not_found_.inc();
+      // Not enough data to read. Wait for it to come.
+      return Network::FilterStatus::StopIteration;
+    }
+    std::string initial_header_buf(initial_header_length, 0);
+    data.copyOut(0, initial_header_length, &initial_header_buf[0]);
```

Comment:
just linearize and use buffer data, no need copyOut

Reply:
Thanks!! Updated.